src/docker.rs

```diff
@@ -589,6 +601,86 @@ impl Docker {
         Ok(docker)
     }
 
+    /// Connect using callback transport.
+    ///
+    /// # Arguments
+    ///
+    ///  - `transport`: transport.
+    ///  - `timeout`: the read/write timeout (seconds) to use for every hyper connection
+    ///  - `client_version`: the client version to communicate with the server.
+    ///
+    /// # Examples
+    ///
+    /// ```rust,no_run
+    /// use bollard::{API_DEFAULT_VERSION, Docker, BollardRequest};
+    /// use futures_util::future::TryFutureExt;
+    /// use futures_util::FutureExt;
+
+    /// let http_connector = hyper_util::client::legacy::connect::HttpConnector::new();
+    ///
+    /// let mut client_builder = hyper_util::client::legacy::Client::builder(hyper_util::rt::TokioExecutor::new());
+    /// client_builder.pool_max_idle_per_host(0);
+    ///
+    /// let client = std::sync::Arc::new(client_builder.build(http_connector));
+    ///
+    /// let connection = Docker::connect_with_callback(
+    ///     Box::new(move |req: BollardRequest| {
+    ///         let client = std::sync::Arc::clone(&client);
+    ///         Box::pin(async move {
+    ///             let (p, b) = req.into_parts();
+    ///             // let _prev = p.headers.insert("host", host);
+    ///             // let mut uri = p.uri.into_parts();
+    ///             //uri.path_and_query = uri.path_and_query.map(|paq|
+    ///             //   uri::PathAndQuery::try_from("/docker".to_owned() + paq.as_str())
+    ///             // ).transpose().map_err(bollard::errors::Error::from)?;
+    ///             // p.uri = uri.try_into().map_err(bollard::errors::Error::from)?;
+    ///             let req = BollardRequest::from_parts(p, b);
+    ///             client.request(req).await.map_err(bollard::errors::Error::from)
+    ///         })
+    ///     }),
+    ///     Some("http://my-custom-docker-server:2735"),
+    ///     4,
+    ///     bollard::API_DEFAULT_VERSION,
+    /// ).unwrap();
+    ///
+    /// connection.ping()
+    ///   .map_ok(|_| Ok::<_, ()>(println!("Connected!")));
+    /// ```
+    pub fn connect_with_callback<S: Into<String>>(
+        callback: CallbackTy,
+        client_addr: Option<S>,
+        timeout: u64,
+        client_version: &ClientVersion,
+    ) -> Result<Docker, Error> {
+        let client_addr = client_addr.map(Into::into).unwrap_or_default();
+        let scheme = client_addr
+            .as_str()
+            .split("://")
+            .next()
+            .map(|s| s.to_owned())
+            .unwrap_or_default();
+        let client_addr = client_addr
+            .trim_start_matches("unix://")
+            .trim_start_matches("npipe://")
+            .trim_start_matches("http://")
+            .trim_start_matches("https://")
+            .trim_start_matches("tcp://")
+            .to_owned();
```

Comment:
This PR looks really good.. I'm wondering here about the `client_addr`, could you unpack it from the `split` that's done in the line above, or remove the characters leading up to `://` ?

Reply:
I changed this to use split_once, tell me if this is what you wanted

Reply:
Looks great!